aggregation/txmetrics/aggregator.go

```diff
@@ -0,0 +1,364 @@
+	out := metricset.Metricset{
+		Timestamp: ts,
+		Metadata: metadata.Metadata{
+			Service: metadata.Service{
+				Name:        key.serviceName,
+				Version:     key.serviceVersion,
+				Environment: key.serviceEnvironment,
+				Agent:       metadata.Agent{Name: key.agentName},
+			},
+			System: metadata.System{
+				DetectedHostname: key.hostname,
+				Container:        metadata.Container{ID: key.containerID},
+				Kubernetes:       metadata.Kubernetes{PodName: key.kubernetesPodName},
+			},
+			User: metadata.User{UserAgent: key.userAgent},
+			// TODO(axw) include client.geo.country_iso_code somewhere
+		},
+		Transaction: metricset.Transaction{
+			Name:   key.transactionName,
+			Type:   key.transactionType,
+			Result: key.transactionResult,
+			Root:   key.traceRoot,
+		},
+		Samples: []metricset.Sample{{
+			Name:   "transaction.duration.histogram",
+			Counts: counts,
+			Values: values,
+		}},
+	}
+	return out
+}
+
+type metrics struct {
+	// TODO(axw) investigate optimising the map for concurrent
+	// updates, along the same lines as what we do in th Go agent
+	// for breakdown metrics.
+	mu      sync.Mutex
+	buckets map[transactionAggregationKey]*transactionMetrics
+}
+
+func newMetrics() *metrics {
+	return &metrics{
+		buckets: make(map[transactionAggregationKey]*transactionMetrics),
+	}
+}
+
+type transactionAggregationKey struct {
+	traceRoot bool
+	agentName string
+	// TODO(axw) requires geoIP lookup in apm-server.
+	//clientCountryISOCode string
+	containerID        string
+	hostname           string
+	kubernetesPodName  string
+	serviceEnvironment string
+	serviceName        string
+	serviceVersion     string
+	transactionName    string
+	transactionResult  string
+	transactionType    string
+	userAgent          string
+}
+
+func makeTransactionAggregationKey(tx *transaction.Event) transactionAggregationKey {
+	deref := func(s *string) string {
+		if s != nil {
+			return *s
+		}
+		return ""
+	}
+	return transactionAggregationKey{
+		traceRoot:         tx.ParentId == nil,
+		transactionName:   deref(tx.Name),
+		transactionResult: deref(tx.Result),
+		transactionType:   tx.Type,
+
+		agentName:          tx.Metadata.Service.Agent.Name,
+		serviceEnvironment: tx.Metadata.Service.Environment,
+		serviceName:        tx.Metadata.Service.Name,
+		serviceVersion:     tx.Metadata.Service.Version,
+
+		hostname:          tx.Metadata.System.Hostname(),
+		containerID:       tx.Metadata.System.Container.ID,
+		kubernetesPodName: tx.Metadata.System.Kubernetes.PodName,
+		userAgent:         tx.Metadata.User.UserAgent,
+
+		// TODO(axw) clientCountryISOCode, requires geoIP lookup in apm-server.
+	}
+}
+
+type transactionMetrics struct {
+	// TODO(axw) consider vendoring or forking hdrhistogram,
+	// since it has been archived in GitHub.
+	histogram *hdrhistogram.Histogram
+}
+
+func (m *transactionMetrics) buckets() (counts []int64, values []float64) {
```

Comment:
The method name `buckets()` is a bit confusing here, as you already use `buckets` for the `map` variable holding `transactionMetrics` (https://github.com/elastic/apm-server/pull/3651/files#diff-2f01ba44a03d252c4b885c9693a9c3c5R284).

Reply:
Noted, will rename this one. Perhaps histogramFields()? Or just distribution()?

Reply:
both SGTM

Reply:
I ended up renaming the other one to "groups" (as in aggregation groups), and this one to "histogramBuckets".